Document sentinel policy support in the public registry (#71)
hashicorp/terraform-docs-common: website/docs/registry/policy-libraries/publishing.mdx

---
page_title: Terraform Registry - Publishing Policy Libraries
description: >-
  How to publish Policy Libraries to the Terraform Registry.
---

# Publishing Policy Libraries

A policy library is a collection of [Sentinel policies](https://docs.hashicorp.com/sentinel/writing) and [policy modules](https://docs.hashicorp.com/sentinel/extending/modules) that anyone can share through the Terraform Registry.

> Policy libraries are built and maintained on GitHub. Once a policy library has been published, you can release a new version of the policies by pushing a properly formatted Git tag.

Reviewer: > properly formatted
Probably we should mention that it must be semver compliant

## Requirements for a Policy Library

- **Public GitHub repository**: The files defining Sentinel policies and policy modules must be stored in a public GitHub repository.

- **Repository name**: GitHub repositories containing Policy Libraries must be named according to the convention `policy-library-<NAME>`, where the `<NAME>` segment provides a name for your policy library.

- **Repository description**: The repository's description is displayed in search and browse pages of the Registry. This should be a short, one sentence description of purpose of the policy library.

> - **sentinel.hcl file**: The GitHub repository must include a `sentinel.hcl` file at its root level. This file outlines the names and file locations for policies and policy modules that are included in the policy library. This is a [Sentinel configuration file](https://docs.hashicorp.com/sentinel/configuration), and it uses the same syntax that a Sentinel author is already familiar with. There are two caveats:

Reviewer: Should we mention that we will ignore all the policies that aren't declared on the `sentinel.hcl` file?

  1. [Remote sources](https://docs.hashicorp.com/sentinel/configuration/remote-sources) for policies and policy modules are not allowed.

>   2. The `enforcement_level` for a policy does not need to be included.

Reviewer: Should we mention that we won't store it at all?

## Publishing to the Registry

### Signing in

Before publishing a policy library, you must first [sign in to the Terraform Registry with a GitHub account](/registry#user-account).

-> If you are publishing a policy library from a repository owned by a GitHub organization that you are a member of, you should verify that the Terraform Registry OAuth app has been granted access to the organization. Go to your [GitHub Settings](https://github.com/settings/applications) and select the Terraform Registry Application under "Authorized OAuth Apps" to confirm.

### Publishing your Policy Library

From the top-right navigation bar of the [Terraform Registry](https://registry.terraform.io), select [Publish > Policy](https://registry.terraform.io/publish/policy) to begin the publishing process. Follow the prompts to select the repository you would like to publish.

You must choose a category for your policy library. Categories are general  themes that can help users find policies that are relevant to their use-case. Choose the category that most closely matches the types of infrastructure to which your policies can be applied.

> You may optionally associate your policy library with a set of [Terraform providers](/registry/providers). You may choose up to 20 publicly-available providers, in order to indicate to users which types of Terraform resources the policies in your library are applicable to.

Reviewer: Just a call out that this number was reduced to 10

---
page_title: Terraform Registry - Publishing Policy Libraries
description: >-
  How to publish Policy Libraries to the Terraform Registry.
---

# Publishing Policy Libraries

A policy library is a collection of [Sentinel policies](https://docs.hashicorp.com/sentinel/writing) and [policy modules](https://docs.hashicorp.com/sentinel/extending/modules) that anyone can share through the Terraform Registry.

Policy libraries are built and maintained on GitHub. Once a policy library has been published, you can release a new version of the policies by pushing a properly formatted Git tag, as described below.

## Requirements for a Policy Library

- **Public GitHub repository**: The files defining Sentinel policies and policy modules must be stored in a public GitHub repository.

> - **Repository name**: GitHub repositories containing Policy Libraries must be named according to the convention `policy-library-<NAME>`, where the `<NAME>` segment provides a name for your policy library.

Reviewer 2: ```suggestion
- **Repository name**: GitHub repositories containing Policy Libraries must be named according to the convention `policy-<NAME>-library`, where the `<NAME>` segment provides a name for your policy library.
```
From what I have experienced, the `<NAME>` segment needs to be between _policy_ and _library_

> - **Public GitHub repository**: The files defining Sentinel policies and policy modules must be stored in a public GitHub repository.
> 
> - **Repository name**: GitHub repositories containing Policy Libraries must be named according to the convention `policy-library-<NAME>`, where the `<NAME>` segment provides a name for your policy library.
> 
> - **Repository description**: The repository's description is displayed in search and browse pages of the Registry. This should be a short, one sentence description of purpose of the policy library.
> 

Reviewer 3: ```suggestion
### GitHub Repository

The Github repository for your policy library must adhere to the following conventions:
- The repository must be public.
- The repository name must follow the convention `policy-library-<NAME>`, where the `<NAME>` segment provides a name for your policy library.
- The repository description should contain a short sentence describing the purpose of the policy library. The Terraform Registry displays this description in search results and on the policy library's details page.
```

- **sentinel.hcl file**: The GitHub repository must include a `sentinel.hcl` file at its root level. This file outlines the names and file locations for policies and policy modules that are included in the policy library. Any policy files in the repository that are not listed in this file will be ignored.

    This is a [Sentinel configuration file](https://docs.hashicorp.com/sentinel/configuration), and it uses the same syntax that a Sentinel author is already familiar with. There are two caveats:

    1. [Remote sources](https://docs.hashicorp.com/sentinel/configuration/remote-sources) for policies and policy modules are not allowed.

>     2. The `enforcement_level` for a policy does not need to be included and is ignored by the Registry.

Author: > It could be nice to have the tiniest little file example here for folks to follow.

Do you think that links to the Sentinel syntax description here is not enough? An example snippet is _very short_ and almost entirely dependent on the context of the library, which gives me some pause. 

Here the `"policy-name"` and path are both at the author's discretion:

```hcl
policy "policy-name" {
  source = "./path/to/somewhere/in/the/repo/policy.sentinel"
}
```

- **`x.y.z` tags for releases**: The registry uses tags to identify versions of policy libraries. Release tag names must be a [semantic version](http://semver.org/), which can optionally be prefixed with a `v`. For example, `v1.0.4` and `0.9.2` are both acceptable. To publish a policy library initially, at least one acceptable tag must be present. Tags that don't match these conventions are ignored.

## Documenting a Policy Library

The Terraform Registry displays documentation for policy libraries, based on Markdown documents that you can choose to include in your GitHub repository.

### Directory Structure

Markdown documents should be in the following locations within the repository. Any Markdown files that are not in these locations will be ignored.

| Location | Filename | Description |
| --- | --- | --- |
| `/` | `README.md` | Index page for the policy library. |
| `docs/policies/` | `<policy name>.md` | Description of a single policy. |
| `docs/modules/` | `<module name>.md` | Description of a single policy module. |

> The `<policy name>` and `<module name>` in the Markdown filenames must match the names of `policy` and `module` blocks in the library's `sentinel.hcl` file.

Author: I was actually surprised to learn that no, the provider documentation is parsed in a different way than documentation for policy libraries and modules. Generally though, the aim is to render markdown files in the registry the same way that they are rendered in GitHub. We expect the user not to be surprised when they move from looking at those files in GitHub to looking at them in the registry.

> The `<policy name>` and `<module name>` in the Markdown filenames must match the names of `policy` and `module` blocks in the library's `sentinel.hcl` file.

Reviewer 3: oh interesting! you know what then? Maybe we add like one sentence after the intro sentence that's like "The registry renders documentation markdown files in a very similar format to how they appear on GitHub." I just wonder if there's a way to be like "don't worry, if it looks good in GitHub it should look fine here! you just need to worry about directory structure:

**Publish new versions, don't edit old versions.** If you have already published a version of your policy library and recognize a mistake or something you would like to change, create a new version of the library. If you instead edit the files in the old version of the library, any users who have been depending on those policies will not be able to use them until they review your changes and update their Sentinel configuration files.

#### Webhooks

Publishing a policy library will create a webhook in your GitHub repository subscribed to the `push` event. When you create new tagged versions of your repository, the webhook will inform the Terraform Registry, which will then be updated to include the new version of your policy library.

If the webhook is missing or not working, you can repair the connection through the Terraform Registry website. From your policy library's index page, select "Manage Policy Library > Resync Library". After a few seconds, a working webhook should be added to your GitHub repository.

> #### Terms of Use
> 
> Anything published to the Terraform Registry is subject to our terms of use. A copy of the terms are available for viewing at <https://registry.terraform.io/terms>

Reviewer 3: Moving this to be part of the Publishing section.
```suggestion
```

> ### Support
> 
> If you experience issues publishing your provider to the Terraform Registry, please contact us at <terraform-registry@hashicorp.com>.

Reviewer 3: ```suggestion
Should we also point them to a repo where they can submit bugs and stuff? or no

## Support

Contact us at <terraform-registry@hashicorp.com> for help publishing your policy library to the Terraform Registry.
```

- The `enforcement_level` for a policy is not required. The Terraform Registry ignores this information if included.

### Version Tags

The Terraform Registry uses tags to identify versions of policy libraries. Version tag names must be a [semantic version](http://semver.org/), which can optionally be prefixed with a `v`. For example, `v1.0.4` and `0.9.2` are both acceptable. To publish a new policy library, at least one acceptable tag must be present. The Terraform Registry ignores tags that do not match these conventions.

## Documenting a Policy Library

The Terraform Registry displays documentation for policy libraries from Markdown files in your GitHub repository.

### Directory Structure

Markdown documents should be in the following locations within the repository. The Terraform Registry ignores Markdown files stored elsewhere.

| Location | Filename | Description |
| --- | --- | --- |
| `/` | `README.md` | Index page for the policy library. |
| `docs/policies/` | `<policy name>.md` | Description of a single policy. |
| `docs/modules/` | `<module name>.md` | Description of a single policy module. |

> The `<policy name>` and `<module name>` in the Markdown filenames must match the names of `policy` and `module` blocks in the library's `sentinel.hcl` file.

Reviewer 3: ```suggestion
### File Names

The `<policy name>` and `<module name>` in the Markdown filenames must match the names of `policy` and `module` blocks in the library's `sentinel.hcl` file.
```

## Publishing to the Registry

> Before you publish, [review the policy library categories on the Terraform Registry](https://registry.terraform.io/browse/policies). Categories are general themes that can help users find policies that are relevant to their use case. We recommend designing your policy libraries to target a specific category, like networking rules or storage security. This helps the community find useful policies more easily.

Reviewer 3: ```suggestion
Before you publish, review the [policy library categories on the Terraform Registry](https://registry.terraform.io/browse/policies). Categories are general themes that can help users find policies that are relevant to their use case. We recommend designing your policy libraries to target a specific category, like networking rules or storage security. This helps the community find useful policies more easily.
```

To publish a policy library to the Terraform Registry:

1. Ensure that your policy library's repository meets the [requirements](#requirements).
1. If a GitHub organization owns the repository, verify that the Terraform Registry OAuth app has access to the organization. Go to your [GitHub Settings](https://github.com/settings/applications) and select the Terraform Registry Application under **Authorized OAuth Apps** to confirm.
1. [Sign in to the Terraform Registry](/registry#user-account) with a GitHub account.

> 1. Select **[Publish > Policy Library](https://registry.terraform.io/publish/policy)**.

Reviewer 3: I just wonder how helpful the link is --> this publish button is readily available on the homepage of the registry, right?

```suggestion
1. Select **Publish > Policy Library**.
```
